src/Aspire.Hosting/Dashboard/KubernetesDataSource.cs

```diff
@@ -0,0 +1,581 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Aspire.Dashboard.Model;
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Dcp;
+using Aspire.Hosting.Dcp.Model;
+using Aspire.Hosting.Dcp.Process;
+using Aspire.Hosting.Utils;
+using k8s;
+using Microsoft.Extensions.Logging;
+
+namespace Aspire.Hosting.Dashboard;
+
+/// <summary>
+/// Pulls data about kubernetes resources from DCP's kubernetes API. Streams updates to consumers.
+/// </summary>
+internal sealed class KubernetesDataSource
+{
+    private readonly KubernetesService _kubernetesService;
+    private readonly DistributedApplicationModel _applicationModel;
+    private readonly Func<ResourceViewModel, ObjectChangeType, ValueTask> _onResourceChanged;
+    private readonly CancellationToken _cancellationToken;
+    private readonly ILogger _logger;
+
+    private readonly Dictionary<string, Container> _containersMap = [];
+    private readonly Dictionary<string, Executable> _executablesMap = [];
+    private readonly Dictionary<string, Service> _servicesMap = [];
+    private readonly Dictionary<string, Endpoint> _endpointsMap = [];
+    private readonly Dictionary<(ResourceKind, string), List<string>> _resourceAssociatedServicesMap = [];
+    private readonly ConcurrentDictionary<string, List<EnvVar>> _dockerEnvironmentByContainerId = [];
+    private readonly HashSet<string> _containerIdsHavingDockerInspections = [];
+
+    public KubernetesDataSource(
+        KubernetesService kubernetesService,
+        DistributedApplicationModel applicationModel,
+        ILoggerFactory loggerFactory,
+        Func<ResourceViewModel, ObjectChangeType, ValueTask> onResourceChanged,
+        CancellationToken cancellationToken)
+    {
+        _kubernetesService = kubernetesService;
+        _applicationModel = applicationModel;
+        _onResourceChanged = onResourceChanged;
+        _cancellationToken = cancellationToken;
+
+        _logger = loggerFactory.CreateLogger<ResourceService>();
+
+        var semaphore = new SemaphoreSlim(1);
```

Comment:
Any particular reason why this code is using `SemaphoreSlim` vs just ordinary lock?

Reply:
So as not to block threads. These are potentially long-running and quite concurrent, especially during startup.

Reply:
Oh, I see, in other words, to wait for it asynchronously. Makes sense.